File: simple-toy.py
import random

def rMaleAge():
    x = random.random()
    if x > 0.86:
        return random.randint(81,85)
    elif x > 0.62:
        return random.randint(76,80)
    elif x > 0.33:
        return random.randint(66,75)
    else:
        return random.randint(50,65)

def rFemaleAge():
    x = random.random()
    if x > 0.93:
        return random.randint(81,85)
    elif x > 0.77:
        return random.randint(76,80)
    elif x > 0.43:
        return random.randint(66,75)
    else:
        return random.randint(50,65)

class RandomPerson(object):
    ANNUAL_RISK_STROKE = [0.0, 0.013, 0.022, 0.032, 0.04, 0.067, 0.098, 0.096, 0.067, 0.152]
    def __init__(self):
        self.gender = "?"
        self.age = -1
        self.nStrokes = 0
    def reset(self):
        x = random.random()
        if x > 0.63:
            self.gender = "F"
            self.age = rFemaleAge()
        else:
            self.gender = "M"
            self.age = rMaleAge()
        x = random.random()
        if x > 0.9:
            self.nStrokes = 2
        elif x > 0.65:
            self.nStrokes = 1
        else:
            self.nStrokes = 0
    def chads(self):
        score = 0
        if self.age >= 75:                   score += 2
        if self.age < 75 and self.age > 65:  score += 1
        if self.gender == "F":               score += 1
        if self.nStrokes > 0:                score += 2
        return score
    def risk(self):
        return self.ANNUAL_RISK_STROKE[self.chads()]
    def alive(self):
        return self.age >= 0
    def dead(self):
        return self.age < 0
    def growOlder(self):
        if random.random() <= self.risk():
            self.nStrokes += 1
            if random.random() <= 0.19:
                self.age = -1 # DEAD
            else: 
                self.age += 1
            return True
        else:
            self.age += 1
            return False

p = RandomPerson()
nStrokes = 0
nDeaths  = 0
for _ in range(50*10000):
    p.reset()
    for _ in range(6):
        if p.growOlder():
            nStrokes += 1
            if p.dead():
                nDeaths += 1
                break

print(nStrokes / 10000.0)
print(nDeaths  / 10000.0)
    

File: simple-toy.R

# Translations of ARENA example
# (aGender == 2)+(aAge >= 75)+
# ((aAge > 65) && aAge < 75))+2*(aStrokeCount >0)
# Note 1: there's minor bug between powerpoint definition
#   and Arena example formula (i.e. it should be age >= 65)
#   this minor bug is kept to test reproducibility
# Note 2: This approach is more readible than Arena
scoreCHADS <- function(gender, age, strokeCount)
{
  score <- 0
  if (age >= 75)              score <- score + 2
  if (age < 75 && age > 65)   score <- score + 1
  if (gender == "F")          score <- score + 1
  if (strokeCount > 0)        score <- score + 2
  return(score)
}

strokeRisk <- function(CHAD)
{
  annualRisk <- c(0, 1.3, 2.2, 3.2, 4.0, 6.7, 9.8, 9.6, 6.7, 15.2)/100
  return(annualRisk[CHAD+1])
}


# Check it out...
# x <- sapply(1:100000, function(x) rMaleAge())
# hist(x, freq=FALSE, breaks=49:86)
rMaleAge <- function()
{
  probs <- c(rep(0.33/16, 16),
             rep(0.29/10, 10),
             rep(0.24/ 5,  5),
             rep(0.14/ 5,  5))
  sample(50:85, 1, prob=probs)
}

rFemaleAge <- function()
{
  probs <- c(rep(0.43/16, 16),
             rep(0.34/10, 10),
             rep(0.16/ 5,  5),
             rep(0.07/ 5,  5))
  sample(50:85, 1, prob=probs)
}

rAge <- function(gender)
{
  if(gender == "M")
  {
    rMaleAge()
  }
  else
  {
    rFemaleAge()
  }
}

rPreviousStroke <- function()
{
  sample(0:2, 1, prob=c(0.65, 0.25, 0.1))
}


asps <- function()
{
  gender      <- sample(c("M","F"), 1, prob=c(0.63, 0.37))
  age         <- rAge(gender)
  prevStrokes <- rPreviousStroke()
  nStrokes    <- prevStrokes
  nEvents     <- 0
  nFatalities <- 0
  
  for(years in 1:6)
  {
    risk      <- strokeRisk(scoreCHADS(gender, age, nStrokes))
    if(runif(1) <= risk)
    {
      nStrokes <- nStrokes + 1
      nEvents  <- nEvents + 1
      if(runif(1) <= 0.19)
      {
        nFatalities <- 1
        break
      }
    }
    age <- age + 1
  }
  
  c(nEvents, nFatalities)
}

sim <- function() rowSums(sapply(1:50, FUN=function(x) asps()))

ptm <- proc.time()
rowSums(sapply(1:10000, function(x) sim()))/10000
proc.time() - ptm